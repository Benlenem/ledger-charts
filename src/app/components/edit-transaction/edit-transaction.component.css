#edit_transaction {
    display:flex;
    flex-direction: column;
}

#header {
    display:flex;
}

#posting {
    display:flex;
}

.expand {
    flex:1;
}

#postings > div > * {
    margin-left: 20px;
}

#header > * {
    margin-left: 20px;
}

.ng-invalid {
    /*
    border:1px solid orange;
    padding:2px;
    */
}

#fab_add {
    bottom: 10px;
    right: 10px;
}

mat-card-content {
    max-height: 300px;
    overflow-y: scroll;
}

mat-card-actions {
    display:flex;
}

#amount {
    text-align: right;
}
