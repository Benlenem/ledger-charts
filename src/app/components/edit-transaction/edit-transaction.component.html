<button *ngIf="!(isPanelOpen | async)" id="fab_add" mat-fab color="accent" (click)="createTransaction()">+</button>

<mat-card *ngIf="isPanelOpen | async" id="edit_transaction">
    <form [formGroup]="group" novalidate (ngSubmit)="onSubmit()">
    <mat-card-content>

    <div id="header">
      <mat-form-field floatPlaceholder="never" id="title" class="expand" >
        <input matInput placeholder="Titre" formControlName="title">
        <mat-error *ngIf="titleErrors | async">
          {{ titleErrors | async }}
        </mat-error>
      </mat-form-field>
  
      <mat-form-field id="date">
          <input matInput [matDatepicker]="picker" placeholder="Date" formControlName="date">
          <mat-datepicker-toggle matSuffix [for]="picker"></mat-datepicker-toggle>
          <mat-datepicker #picker></mat-datepicker>
          <mat-error *ngIf="dateErrors | async">
            {{ dateErrors | async }}
          </mat-error>
      </mat-form-field>

      <button type="button" id="close" (click)="closePanel()" mat-icon-button>
          <mat-icon class="material-icons">clear</mat-icon>
      </button>
    </div>

    <div formArrayName="tags">
        <app-tags
        [tags]="group.controls.tags.value"
        [allowEditing]="true"
        (onClosedClick)="removeHeaderTag($event)"
        (onTagAdded)="addHeaderTag($event)"
        ></app-tags>
    </div>

    <div id="postings" formArrayName="postings" [ngClass]="{hasErrors:(postings.errors)}">
      <div *ngFor="let p of postings.controls; let i=index" [formGroupName]="i" class="posting" >
        <mat-form-field class="expand" floatPlaceholder="never">
          <input matInput placeholder="Compte" formControlName="account" type="text" [matAutocomplete]="auto">
          <mat-error *ngIf="postingAccountErrors(i) | async">
            {{ postingAccountErrors(i) | async }}
          </mat-error>
        </mat-form-field>

        <mat-form-field id="amount" floatPlaceholder="never">
            <input matInput placeholder="Montant" formControlName="amount">
            <mat-error *ngIf="postingAmountErrors(i) | async">
              {{ postingAmountErrors(i) | async }}
            </mat-error>
        </mat-form-field>

        <mat-form-field floatPlaceholder="never">
          <input matInput placeholder="Commentaire" formControlName="comment">
          <mat-error *ngIf="postingCommentErrors(i) | async">
            {{ postingCommentErrors(i) | async }}
          </mat-error>
        </mat-form-field>

        <button type="button" id="close" (click)="removePosting(i)" mat-icon-button>
          <mat-icon class="material-icons">delete</mat-icon>
        </button>
      </div>

      <mat-error *ngIf="postingsErrors | async">
        {{ postingsErrors | async }}
      </mat-error>
    </div>

    <mat-autocomplete #auto="matAutocomplete">
      <mat-option *ngFor="let account of filteredAccounts | async" [value]="account.name">
          {{ account.name }}
      </mat-option>
    </mat-autocomplete>

  </mat-card-content>

  <mat-card-actions>
    <button mat-raised-button type="submit" [disabled]="group.pristine || group.invalid" color="primary">Sauvegarder</button>
    <button mat-raised-button type="button" (click)="revert()" [disabled]="group.pristine">RÃ©initialiser</button>
    <button mat-raised-button type="menu" (click)="addPosting();$event.preventDefault();">Ajouter une ligne</button>
    <button mat-raised-button type="menu" (click)="deleteTransaction();$event.preventDefault();" *ngIf="transactionToEdit" color="warn">Supprimer la transaction</button>
  </mat-card-actions>

</form>

</mat-card>
