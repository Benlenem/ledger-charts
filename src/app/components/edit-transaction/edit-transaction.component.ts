import { NgRedux } from '@angular-redux/store';
import { Component, OnInit, ViewChild } from '@angular/core';
import {
  AbstractControl, AsyncValidatorFn, FormArray, FormBuilder, FormControl, FormGroup, ValidationErrors,
  Validators,
} from '@angular/forms';
import { MatDatepicker } from '@angular/material';

import Decimal from 'decimal.js';
import { Observable } from 'rxjs/Observable';
import { Subject } from 'rxjs/Subject';
import { Transaction, TransactionWithUUID, isTransactionWithUUID } from '../../shared/models/transaction';

import * as moment from 'moment';
import 'rxjs/add/observable/concat';
import 'rxjs/add/observable/of';
import 'rxjs/add/operator/do';
import 'rxjs/add/operator/mergeMap';
import { Posting } from '../../shared/models/posting';
import { AppState } from '../../shared/models/app-state';
import {
  selectEditedTransaction,
  AppStateActions,
  allAccountsSelector,
  isTransactionPanelOpenSelector,
} from '../../shared/reducers/app-state-reducer';
import { Account } from '../../shared/models/account';

@Component({
  selector: 'app-edit-transaction',
  styleUrls: ['./edit-transaction.component.css'],
  templateUrl: './edit-transaction.component.html',
})
export class EditTransactionComponent implements OnInit {

  @ViewChild(MatDatepicker) myDatepicker: MatDatepicker<Date>;

  transactionToEdit?: TransactionWithUUID | Transaction;
  isPanelOpen: Observable<boolean>;
  group: FormGroup;
  filteredAccounts: Subject<Account[]> = new Subject();
  formErrors: Observable<string>;

  constructor(private ngRedux: NgRedux<AppState>, private _formBuilder: FormBuilder) { }

  ngOnInit() {
    this.ngRedux.select(selectEditedTransaction).subscribe( tr => {
      this.transactionToEdit = tr;
      this.init();
    });

    this.isPanelOpen = this.ngRedux.select(isTransactionPanelOpenSelector);
  }

  createTransaction() {
    this.ngRedux.dispatch(AppStateActions.setEditedTransaction({
      header: {
        date: moment.utc().unix(),
        title: '',
      },
      postings: [],
    }));
    this.init();
  }

  createPostingGroup(): FormGroup {
    const accountFormControl = new FormControl('', Validators.required);

    accountFormControl.valueChanges
    .flatMap(v => this.filterAccount(v))
    .subscribe( accounts => this.filteredAccounts.next(accounts));

    return this._formBuilder.group({
      account: accountFormControl,
      amount: [null, Validators.pattern(/^-?\d+(\.\d+)?$/)],
      comment: [''],
      currency: [''],
    });
  }
  private init() {
      // Prepare form structure
      this.group = this._formBuilder.group({
        date: ['', [Validators.required]],
        postings: this._formBuilder.array([]),
        title: ['', [Validators.required]],
      });

      const postingsControl = this.transactionToEdit ? this.transactionToEdit.postings.map(p => this.createPostingGroup()) : [];
      this.group.setControl('postings', this._formBuilder.array(postingsControl, null, postingsRepartitionAsyncValidator()));

      // Initialize values
      const title = this.transactionToEdit ? this.transactionToEdit.header.title : '';
      const date = this.transactionToEdit ? moment.unix(this.transactionToEdit.header.date) : moment();
      const postings = this.transactionToEdit ? this.transactionToEdit.postings.map( p => {
        return {
          account: p.account,
          amount: p.amount ? p.amount.toString() : null,
          comment: p.comment || '',
          currency: p.currency || '',
        };
      }) : [];

      // Set all the values
      this.group.setValue({
        date: date,
        postings: postings,
        title: title,
      });

      const errorObservable = Observable.of(this.logErrors(this.group));
      const groupErrors = this.group.valueChanges.map( t => this.logErrors(this.group));

    this.formErrors = Observable.concat(errorObservable, groupErrors)
      .map(e => e[0]);

      this.group.markAsDirty();
      this.group.updateValueAndValidity();
  }

  get postings(): FormArray {
    return this.group.get('postings') as FormArray;
  }

  filterAccount(query: any): Observable<Account[]> {
    return this.ngRedux.select(allAccountsSelector).map(accounts => {
      return Object.values(accounts).filter(a => a.name.toLowerCase().indexOf(query.toLowerCase()) !== -1);
    });
  }

  onSubmit() {
    const tr = this.prepareTransaction();

    if (!this.transactionToEdit) { return; }

    if (isTransactionWithUUID(this.transactionToEdit)) {
      const modifiedTransaction = {
        ...tr,
        uuid: this.transactionToEdit.uuid,
      };
      this.ngRedux.dispatch(AppStateActions.updateTransaction(modifiedTransaction));
    } else {
      this.ngRedux.dispatch(AppStateActions.addTransactions([tr]));
    }

    this.closePanel();
  }

  revert() {
    this.init();
  }

  private prepareTransaction(): Transaction {
    const formModel = this.group.value;

    return {
      header: {
        date: formModel.date.unix(),
        tag: '',
        title: formModel.title,
      },
      postings: formModel.postings.map(p => {
        return {
          account: p.account,
          amount: p.amount,
          comment: p.comment,
          currency: p.currency,
          tag: '',
        } as Posting;
      }),
    };
  }

  logErrors(control: AbstractControl, id: string = ''): string[] {
    const e = control.errors;
    let errors: string[] = [];

    if (e != null) {
      errors.push(id + ': ' + JSON.stringify(e));
    }

    if (control instanceof FormGroup) {
      Object.keys(control.controls).forEach(k => {
        const c = control.get(k);
        if (c) {
          errors = errors.concat(this.logErrors(c, id + '/' + k));
        }
      });
    } else if (control instanceof FormArray) {
      control.controls.forEach((c, i) => errors = errors.concat(this.logErrors(c, id + '[' + i + ']')));
    }

    return errors;
  }

  addPosting() {
    const postings = (this.group.get('postings') as FormArray);
    postings.push(this.createPostingGroup());
    postings.markAsDirty();
  }

  removePosting(index: number) {
    const postings = (this.group.get('postings') as FormArray);
    postings.removeAt(index);
    postings.markAsDirty();
  }

  deleteTransaction() {
    if (isTransactionWithUUID(this.transactionToEdit)) {
      this.ngRedux.dispatch(AppStateActions.deleteTransaction(this.transactionToEdit.uuid));
    }
  }

  closePanel() {
    this.ngRedux.dispatch(AppStateActions.setEditedTransaction(undefined));
  }
}

export function postingsRepartitionAsyncValidator(): AsyncValidatorFn {
  return (array: FormArray): Observable<ValidationErrors | null> => {
    const accountControls = array.controls
      .map(c => c.get('account'))
      .filter(ac => ac != null)
      // tslint:disable-next-line no-non-null-assertion
      .map(ac => ac!.value as string);
    const accounts = new Set(accountControls);

    let error: ValidationErrors | null = null;
    if (accounts.size < 2) {
      error = { 'notEnoughAccounts': 'minimum is 2'};
    } else {
      const amountControls = array.controls
        .map(c => c.get('amount'))
        .filter(ac => ac != null)
        // tslint:disable-next-line no-non-null-assertion
        .map(ac => ac!.value as string);

      const howManyNulls = amountControls.filter(a => a == null || a.trim() === '').length;

      if (howManyNulls > 1) {
        error = { 'onlyOneNullAmount': null };
      } else if (howManyNulls === 0) {
          const sum = amountControls.filter(a => a != null && a.trim() !== '').map(a => Number.parseFloat(a)).reduce((p, c) => p + c, 0);
          if (sum !== 0) {
            error = { 'incorrectBalance': sum };
          }
      }
    }
    return Observable.of(error);
  };
}
