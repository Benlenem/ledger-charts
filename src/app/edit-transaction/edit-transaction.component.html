<button *ngIf="!isEditing" id="fab_add" md-fab color="accent" (click)="createTransaction()">+</button>

<md-card *ngIf="isEditing" id="edit_transaction">
    <form [formGroup]="group" novalidate (ngSubmit)="onSubmit()">
    <md-card-content>
  
    <div id="header">
      <md-input-container floatPlaceholder="never" id="title" class="expand">
        <input mdInput #filter placeholder="Titre" formControlName="title">
      </md-input-container>
      
      <md-input-container id="date">
          <input mdInput [mdDatepicker]="picker" placeholder="Date" formControlName="date">
          <md-datepicker-toggle mdSuffix [for]="picker"></md-datepicker-toggle>
        </md-input-container>
        <md-datepicker #picker></md-datepicker>

        <button id="close" (click)="closePanel()" md-icon-button>
            <md-icon class="material-icons">clear</md-icon>
          </button>
    </div>

    <div id="postings" formArrayName="postings">
      <div *ngFor="let p of postings.controls; let i=index" [formGroupName]="i" id="posting" >
        <md-input-container class="expand" floatPlaceholder="never">
          <input mdInput placeholder="Compte" formControlName="account" type="text" [mdAutocomplete]="auto">
        </md-input-container>
        <md-input-container class="expand" floatPlaceholder="never">
          <input mdInput #filter placeholder="Commentaire" formControlName="comment">
        </md-input-container>
        <md-input-container id="amount" floatPlaceholder="never">
            <input mdInput #filter placeholder="Montant" type="number" formControlName="amount">
        </md-input-container>
        <md-input-container id="currency" floatPlaceholder="never">
            <input mdInput #filter placeholder="Monnaie" formControlName="currency">
        </md-input-container>
        <button id="close" (click)="removePosting(i)" md-icon-button>
          <md-icon class="material-icons">delete</md-icon>
        </button>
      </div>
    </div>

    <md-autocomplete #auto="mdAutocomplete">
      <md-option *ngFor="let account of filteredAccounts | async" [value]="account.name">
          {{ account.name }}
      </md-option>
    </md-autocomplete>
       
  </md-card-content>

  <md-card-actions>
      <button md-raised-button type="submit" [disabled]="group.pristine || group.invalid" color="primary">Sauvegarder</button>
      <button md-raised-button type="button" (click)="revert()" [disabled]="group.pristine">RÃ©initialiser</button>
    
    <button md-raised-button type="menu" (click)="addPosting();$event.preventDefault();">Ajouter une ligne</button>
    <button md-raised-button type="menu" (click)="deleteTransaction();$event.preventDefault();" *ngIf="transaction?.uuid" color="warn">Supprimer la transaction</button>

    <div class="expand"></div>

    <button md-raised-button *ngIf="group.status === 'INVALID'" type="button" color="warn"><md-icon class="material-icons">warning</md-icon>{{ (formErrors | async) }}</button>

    
  
  </md-card-actions>

</form>

</md-card>