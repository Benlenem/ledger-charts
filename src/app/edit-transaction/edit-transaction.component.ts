import { Component, OnInit, Input, ViewChild } from '@angular/core'
import { JsonPipe } from '@angular/common';
import {MdDatepicker} from '@angular/material'
import { FormControl, FormGroup, FormBuilder, Validators, FormArray, ValidatorFn, AsyncValidatorFn, AbstractControl, ValidationErrors }            from '@angular/forms';

import {Observable, Subject} from 'rxjs'
import { Transaction, Posting } from '../models/models'
import { AppStateService } from '../app-state.service'

import * as moment from "moment"

@Component({
  selector: 'edit-transaction',
  templateUrl: './edit-transaction.component.html',
  styleUrls: ['./edit-transaction.component.css']
})
export class EditTransactionComponent implements OnInit {

  @ViewChild(MdDatepicker) myDatepicker: MdDatepicker<Date>;
  
  private transaction: Transaction | undefined

  isEditing: boolean
  group : FormGroup
  filteredAccounts: Subject<Account[]> = new Subject()
  formErrors: Observable<string>

  constructor(private _state: AppStateService, private _formBuilder: FormBuilder) { }

  ngOnInit(){
    this._state.editedTransaction().subscribe( tr => {
      this.transaction = tr
      this.isEditing = this.transaction != undefined
      this.init()
    });
  }

  createTransaction(){
    this.isEditing = true
    this.init()
  }

  createPostingGroup(): FormGroup{
    let accountFormControl = new FormControl('', Validators.required)

    accountFormControl.valueChanges
    .flatMap(v => this.filterAccount(v))
    .subscribe( accounts => this.filteredAccounts.next(accounts))

    return this._formBuilder.group({
      account: accountFormControl,
      amount: [null, Validators.pattern(/^-?\d+(\.\d+)?$/)],
      currency: [''],
      comment: ['']
    })
  }
  private init(){
      //Prepare form structure
      this.group = this._formBuilder.group({
        title: ['', [Validators.required]],
        date: ['', [Validators.required]],
        postings: this._formBuilder.array([])
      });
      
      let postingsControl = this.transaction ? this.transaction.postings.map(p => this.createPostingGroup()) : []
      this.group.setControl('postings', this._formBuilder.array(postingsControl, null, postingsRepartitionAsyncValidator()))

      //Initialize values
      let title = this.transaction ? this.transaction.header.title : ''
      let date = this.transaction ? this.transaction.header.date : moment()
      let postings = this.transaction ? this.transaction.postings.map( p => {
        return { 
          account: p.account,
          amount: p.amount ? p.amount.toString() : null,
          currency: p.currency,
          comment: p.comment
        }
      }) : []

      //Set all the values
      this.group.setValue({
        title: title,
        date: date,
        postings: postings
      })

      this.formErrors = Observable
      .of(this.logErrors(this.group))
      .concat(this.group.valueChanges.map( t => this.logErrors(this.group)))
      .do(e => console.log(e))
      .map(e => e[0])
      
      this.group.markAsDirty()
      this.group.updateValueAndValidity()
  }

  get postings(): FormArray {
    return this.group.get('postings') as FormArray;
  };

  filterAccount(query: any) : Observable<Account[]>{
    return this._state.allAccountsFlattened().map(accounts => {
      return accounts.filter(a => a.name.toLowerCase().indexOf(query.toLowerCase()) != -1)
    })
  }

  onSubmit() {
    let tr = this.prepareTransaction()
    this._state.createOrUpdateTransaction(tr).subscribe()
    this.closePanel()
  }

  revert() {
    this.init()
  }

  private prepareTransaction(): Transaction {
    const formModel = this.group.value
  
    let transaction : Transaction= {
      uuid: this.transaction ? this.transaction.uuid : undefined,
      header: {
        date: formModel.date,
        title: formModel.title,
        tag: ''
      },
      postings: formModel.postings.map(p => {
        return {
          tag: '',
          account: p.account,
          amount: p.amount ? Number.parseFloat(p.amount) : null,
          comment: p.comment,
          currency: p.currency
        } as Posting
      })
    }

    return transaction
  }

  logErrors(c: AbstractControl, id: string = "") : string[] {
    let e = c.errors
    let errors : string[] = []

    if(e != null){
      errors.push(id + ": " + JSON.stringify(e))
    }

    if(c instanceof FormGroup){ 
      Object.keys(c.controls).forEach(k => errors = errors.concat(this.logErrors(c.get(k)!, id+"/"+k)))
    }
    else if(c instanceof FormArray){
      c.controls.forEach((c, i) => errors = errors.concat(this.logErrors(c, id+"["+i+"]")))
    }

    return errors;
  }

  addPosting() {
    let postings = (this.group.get('postings') as FormArray)
    postings.push(this.createPostingGroup())
    postings.markAsDirty()
  }

  removePosting(index: number){
    let postings = (this.group.get('postings') as FormArray)
    postings.removeAt(index)
    postings.markAsDirty()
  }

  deleteTransaction(){
    this._state.deleteTransaction(this.transaction).subscribe()
  }

  closePanel(){
    this._state.setEditedTransaction(undefined).subscribe()
  }
}

export function postingsRepartitionAsyncValidator(): AsyncValidatorFn {
  return (array: FormArray): Observable<ValidationErrors | null> => {
    console.log("async")
    let accounts = new Set(array.controls.map(c => c.get('account')!.value as string).filter(a => a != ""))

    let error :ValidationErrors | null = null; 
    if(accounts.size < 2) {
      error = { 'notEnoughAccounts': 'minimum is 2'}
    }
    else {
      let amounts = array.controls.map(c => c.get('amount')!.value as string)      
      let howManyNulls = amounts.filter(a => a == null || a.trim() == '').length

      if(howManyNulls > 1){
        error = { 'onlyOneNullAmount': null }
      }
      else if(howManyNulls == 0) {
          let sum = amounts.filter(a => a != null && a.trim() != '').map(a => Number.parseFloat(a)).reduce((p, c) => p + c, 0)
          if(sum != 0){
            error = { 'incorrectBalance': sum }
          }
      }
    }
    return Observable.of(error)
  };
}