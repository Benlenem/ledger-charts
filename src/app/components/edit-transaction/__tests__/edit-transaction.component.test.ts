import { postingsRepartitionAsyncValidator } from '../edit-transaction.component';
import { FormBuilder, FormArray, FormGroup } from '@angular/forms';

const createPostingGroup = (b: FormBuilder): FormGroup => {
  return b.group({
    account: [''],
    amount: [null],
    comment: [''],
    currency: [''],
  });
};

const createPostingGroupWithValue = (b: FormBuilder, account: string, amount: string | null, comment: string, currency: string) => {
  const g = createPostingGroup(b);
  g.setValue({account, amount, comment, currency});
  return g;
};

describe(postingsRepartitionAsyncValidator.name, () => {
  it('returns error when no postings', (done) => {
    const postings = new FormArray([]);
    const res = postingsRepartitionAsyncValidator(postings).toPromise();

    res.then( errors => {
      expect(errors).toMatchSnapshot();
      done();
    });
  });

  it('returns error when one empty posting', (done) => {
    const b = new FormBuilder();
    const postingGroup = createPostingGroup(b);
    const postings = new FormArray([postingGroup]);

    const res = postingsRepartitionAsyncValidator(postings).toPromise();

    res.then( errors => {
      expect(errors).toMatchSnapshot();
      done();
    });
  });

  it('returns error when more than one empty amount', (done) => {
    const b = new FormBuilder();
    const p1 = createPostingGroupWithValue(b, 'Expenses', null, '', '');
    const p2 = createPostingGroupWithValue(b, 'Income', ' ', '', '');

    const postings = new FormArray([p1, p2]);

    const res = postingsRepartitionAsyncValidator(postings).toPromise();

    res.then( errors => {
      expect(errors).toMatchSnapshot();
      done();
    });
  });

  it('returns error when integer balance is incorrect', (done) => {
    const b = new FormBuilder();
    const p1 = createPostingGroupWithValue(b, 'Expenses', '3.14', '', '');
    const p2 = createPostingGroupWithValue(b, 'Income', '-3.15', '', '');

    const postings = new FormArray([p1, p2]);

    const res = postingsRepartitionAsyncValidator(postings).toPromise();

    res.then( errors => {
      expect(errors).toMatchSnapshot();
      done();
    });
  });

  it('returns no error with one amount set', (done) => {
    const b = new FormBuilder();
    const p1 = createPostingGroupWithValue(b, 'Expenses', '3.14', '', '');
    const p2 = createPostingGroupWithValue(b, 'Income', null, '', '');

    const postings = new FormArray([p1, p2]);

    const res = postingsRepartitionAsyncValidator(postings).toPromise();

    res.then( errors => {
      expect(errors).toBeNull();
      done();
    });
  });

  it('returns no error with all amounts sets', (done) => {
    const b = new FormBuilder();
    const p1 = createPostingGroupWithValue(b, 'Expenses', '3.14', '', '');
    const p2 = createPostingGroupWithValue(b, 'Income', '-3.14', '', '');

    const postings = new FormArray([p1, p2]);

    const res = postingsRepartitionAsyncValidator(postings).toPromise();

    res.then( errors => {
      expect(errors).toBeNull();
      done();
    });
  });

  // See http://0.30000000000000004.com/ and https://en.wikipedia.org/wiki/Floating_point
  it('returns no error even with float imprecision', (done) => {
    const b = new FormBuilder();
    const p1 = createPostingGroupWithValue(b, 'Expenses', '0.1', '', '');
    const p2 = createPostingGroupWithValue(b, 'Income', '0.2', '', '');
    const p3 = createPostingGroupWithValue(b, 'Income', '-0.3', '', '');

    const postings = new FormArray([p1, p2, p3]);

    const res = postingsRepartitionAsyncValidator(postings).toPromise();

    res.then( errors => {
      expect(errors).toBeNull();
      done();
    });
  });
});
