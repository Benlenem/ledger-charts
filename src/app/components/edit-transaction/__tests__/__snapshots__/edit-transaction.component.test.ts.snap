// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns error when integer balance is incorrect 1`] = `
Object {
  "incorrectBalance": "-0.01",
}
`;

exports[`returns error when more than one empty amount 1`] = `
Object {
  "onlyOneNullAmount": null,
}
`;

exports[`returns error when no postings 1`] = `
Object {
  "notEnoughAccounts": "minimum is 2",
}
`;

exports[`returns error when one empty posting 1`] = `
Object {
  "notEnoughAccounts": "minimum is 2",
}
`;
